Packages/com.unity.inputsystem/InputSystem/Plugins/HID/HIDSupport.cs

```diff
@@ -50,6 +62,8 @@ public static void Initialize()
             #endif
         }
 
+        public static Dictionary<HID.UsagePage, HashSet<int>> supportedUsages { get; set; }
```

Comment:
Hmm, the dictionary pains me a little :) A dictionary with a nested hashset for the case here... that's four heap objects for a relatively trivial lookup. Why not just a flat array of tuples?

Looking at the API, I'm also wondering now whether a callback wouldn't be better. Just gives the whole decision in the user's hands instead of just having a whitelist.

```
    using ShouldCreateHIDCallback = Func<HID.HIDDeviceDescriptor, bool?>;

//...

public static class HIDSupport
{
        public static event ShouldCreateHIDCallback shouldCreateHID
        {
            add => s_ShouldCreateHID.Append(value);
            remove => s_ShouldCreateHID.Remove(value);
        }

        internal static InlinedArray<ShouldCreateHIDCallback> s_ShouldCreateHID;

        private static bool? DefaultShouldCreateHIDCallback(HID.HIDDeviceDescriptor descriptor)
        {
            if (descriptor.usagePage == HID.UsagePage.GenericDesktop)
            {
                switch (descriptor.usage)
                {
                    case (int)HID.GenericDesktop.Joystick:
                    case (int)HID.GenericDesktop.Gamepad:
                    case (int)HID.GenericDesktop.MultiAxisController:
                        return true;
                }
            }
            return null;
        }
        
        /// <summary>
        /// Add support for generic HIDs to InputSystem.
        /// </summary>
        public static void Initialize()
        {
            s_ShouldCreateHID.Append(DefaultShouldCreateHIDCallback);

            //...
```

And then in the HID class.

```
            // Check callbacks to see whether we should actually create a device for this specific HID.
            // If no callback says yes or one says no, we ignore the device.
            if (HIDSupport.s_ShouldCreateHID.All(f => f(hidDeviceDescriptor) != true) ||
                HIDSupport.s_ShouldCreateHID.Any(f => f(hidDeviceDescriptor) == false))
                return null;
```

Seems more flexible. But also a bit more code.

What do you think?

Reply:
I like it. (As i wrote, wasn't really sure about the Dictionary/HashSet myself, that felt kind of heavy).